### PR TITLE
Initial addition of the Embedded Rust Book

### PR DESCRIPTION
**Pull request by jamesmunns**
This PR adds the Embedded Rust Book to the bookshelf as a submodule, and adds text for the bookshelf page. I have added a new section after "Master Rust" called "Specialize Rust", with the plan that future domain WG books can also reside here. This now extends the titles down to H3, where formerly only H1 and H2 were used.

The added submodule tracks the master branch of the Embedded WG repo.

If there are additional steps necessary to make this work in CI (perhaps adding this to `src/ci/docker/x86_64-gnu-tools/checktools.sh:32` or so?), please let me know.

CC @steveklabnik @japaric 

Also CC issue https://github.com/rust-embedded/wg/issues/257

### CODE REVIEW

**Comment by jamesmunns** [comment]
One of our CI scripts was exactly one line too long. Moving to a WIP branch to resolve any issues. Please do not merge until I have moved the submodule back to tracking "master".

**Comment by steveklabnik** [comment]
This looks great!

> If there are additional steps necessary to make this work in CI (perhaps adding this to src/ci/docker/x86_64-gnu-tools/checktools.sh:32 or so?), please let me know.

Yes, you'll need to modify rustbuild as well. It *should* be as easy as adding a line in here https://github.com/rust-lang/rust/blob/master/src/bootstrap/doc.rs#L70-L76

**Comment by steveklabnik** [comment]
I thought that mdbook always put the first page of the book as index.html; maybe I'm wrong...

**Comment by jamesmunns** [comment]
Hmm, you seem to be right @steveklabnik:

```
➜  book git:(master) rm -rf ./book
➜  book git:(master) mdbook build
2018-11-28 15:30:59 [INFO] (mdbook::book): Book building has started
2018-11-28 15:30:59 [INFO] (mdbook::book): Running the html backend
➜  book git:(master) ls -hal ./book/index.html
-rw-r--r-- 1 james james 19K Nov 28 15:30 ./book/index.html
```

I guess this means I have some investigating to do, support would be appreciated, otherwise I'll look once I have a chance.

**Comment by steveklabnik** [comment]
[01:35:28] index.html:103: broken link - embedded-book/index.html

This means that `index.html` is linking to `embedded-book/index.html`. are you sure that this doesn't mean that *your* index isn't trying to link to a page deeper than itself?

**Comment by steveklabnik** [comment]
I cannot get this to build due to a really strange error

```
running: "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.15.26726\\bin\\HostX64\\x64\\cl.exe" "/nologo" "/MT"
"/O2" "/Zl" "/D__func__=__FUNCTION__" "/FoC:\\Users\\steve\\src\\rust\\build\\x86_64-pc-windows-msvc\\stage0-std\\x86_64-pc-windows-msvc\\release\\bui
ld\\compiler_builtins-5994e1b3add468a9\\out\\../../libcompiler_builtins\\compiler-rt/lib/builtins\\absvdi2.o" "/c" "../../libcompiler_builtins\\compil
er-rt/lib/builtins\\absvdi2.c"
absvdi2.c
c1: fatal error C1083: Cannot open source file: '../../libcompiler_builtins\compiler-rt/lib/builtins\absvdi2.c': No such file or directory
exit code: 2
```

**Review by steveklabnik — CHANGES_REQUESTED** [review]
Sooooooooooooo @jamesmunns , I forgot a crucial step:

```diff
diff --git a/src/bootstrap/builder.rs b/src/bootstrap/builder.rs
index 079fdee7e8..97d3ecba04 100644
--- a/src/bootstrap/builder.rs
+++ b/src/bootstrap/builder.rs
@@ -442,7 +442,8 @@ impl<'a> Builder<'a> {
                 doc::RustdocBook,
                 doc::RustByExample,
                 doc::RustcBook,
-                doc::CargoBook
+                doc::CargoBook,
+                doc::EmbeddedBook
             ),
             Kind::Dist => describe!(
                 dist::Docs,
```

the changes you have *define* the build, but don't actually do the building.

after that, we get this report:

```text
embedded-book\intro\install\linux.html:237: broken link - embedded-book\verify.md
embedded-book\intro\install\macos.html:147: broken link - embedded-book\verify.md
embedded-book\intro\install\verify.html:170: broken link - hardware.md
embedded-book\intro\install\verify.html:177: broken link - hardware.md
embedded-book\intro\install\verify.html:180: broken link - embedded-book\linux.md
embedded-book\intro\install\windows.html:161: broken link - embedded-book\verify.md
embedded-book\intro\install.html:160: broken link - embedded-book\install\linux.md
embedded-book\intro\install.html:161: broken link - embedded-book\install\windows.md
embedded-book\intro\install.html:162: broken link - embedded-book\install\macos.md
embedded-book\print.html:374: broken link - embedded-book\install\linux.md
embedded-book\print.html:375: broken link - embedded-book\install\windows.md
embedded-book\print.html:376: broken link - embedded-book\install\macos.md
embedded-book\print.html:479: broken link - embedded-book\verify.md
embedded-book\print.html:491: broken link - embedded-book\verify.md
embedded-book\print.html:517: broken link - embedded-book\verify.md
embedded-book\print.html:552: broken link - hardware.md
embedded-book\print.html:559: broken link - hardware.md
embedded-book\print.html:562: broken link - embedded-book\linux.md
embedded-book\print.html:628: broken link - embedded-book\hardware.md
embedded-book\print.html:1107: broken link - embedded-book\qemu.md
embedded-book\print.html:1151: broken link - intro\install\verify.md
embedded-book\print.html:1173: broken link - intro\install\verify.md
embedded-book\print.html:1329: broken link - portability\index.md
embedded-book\print.html:1352: broken link - peripherals\index.md
embedded-book\start\hardware.html:177: broken link - embedded-book\qemu.md
embedded-book\start\hardware.html:221: broken link - intro\install\verify.md
embedded-book\start\hardware.html:243: broken link - intro\install\verify.md
embedded-book\start\index.html:146: broken link - embedded-book\hardware.md
embedded-book\start\registers.html:142: broken link - portability\index.md
embedded-book\start\registers.html:165: broken link - peripherals\index.md
```

all of the `print` ones are duplicates of the regular pages, so this list is only half as long.

make sense?

**Comment by jamesmunns** [comment]
I'll take a look tomorrow, I've got some time on the train. I'll see what I can get together :)

**Comment by Dylan-DPC** [comment]
ping from triage @jamesmunns you need to update this with the changes requested.

**Comment by jamesmunns** [comment]
Thanks @dylan-dpc, I still intend to finish this as soon as I can.

**Comment by jamesmunns** [comment]
@steveklabnik I think I know what the issue is, the main rust repo uses an older version of `mdbook` than the embedded wg (`0.1.7` vs `0.2.2`). We rely on using links to relative `md` files within the pre-rendered markdown, and newer versions of `mdbook` automatically render this to `.html` links.

In [this change](https://github.com/rust-lang/rust/pull/56291/commits/b115a5535c8b7a270a4ee2e0ce93810874ddf773#diff-612b38fc566f4724cd224c7b21349ab5R11), I update the dependency to `0.2.2`. After updating, `x.py dist` still seems to complete successfully.

Let's see how the CI goes.

**Comment by steveklabnik** [comment]
Updating to 0.2 is going to take a massive effort across all books. I plan on trying to tackle it soon, but it’s tough.

**Comment by jamesmunns** [comment]
Oh. That's unfortunate. Let me ask the folks at the next embedded meeting whether we are okay with breaking convention on how relative links should be formatted.

@steveklabnik how would I run exactly the same test that CI is running for this check? I don't get the same errors when I just run `x.py dist` or `x.py doc`, I think it's missing the linkcheck step?

I might be interested in chipping away at the mdbook 0.2 changes, if they are mostly mechanical.

**Comment by steveklabnik** [comment]
python x.py test src/tools/linkchecker

It’s mostly about the re-basing of relative URLs, as you mentioned. The book, the reference, the nomicon, RBE... they’ll all need to be updated.

**Comment by jamesmunns** [comment]
Okay, let me take a pass at this, and see how much damage I can do.

https://github.com/rust-embedded/book/issues/55 is relevant, it's where the embedded wg decided on a canonical link format (using mdbook 0.2).

**Comment by jamesmunns** [comment]
Hey @steveklabnik, I had a bit of an idea. Would you be up for supporting multiple versions of mdbook at the same time? I think this could solve our problem today, and it could make it easier to incrementally port books between versions.

Take a look at [this commit](https://github.com/rust-lang/rust/pull/56291/commits/6c60e3a026540e381fa5ace6edec6a760568b0ff), particularly:

* [Cargo.toml trickery](https://github.com/rust-lang/rust/blob/6c60e3a026540e381fa5ace6edec6a760568b0ff/src/tools/rustbook/Cargo.toml#L13-L23),
* [Handling two render steps](https://github.com/rust-lang/rust/blob/6c60e3a026540e381fa5ace6edec6a760568b0ff/src/tools/rustbook/src/main.rs#L48-L69)
* [Describing the version to be used](https://github.com/rust-lang/rust/blob/6c60e3a026540e381fa5ace6edec6a760568b0ff/src/bootstrap/doc.rs#L72-L80)

There are definitely some cleanups to be made, and the `print` links are still failing for some reason, but I wanted to check with you to see if this is a route you would be willing to take.

**Comment by jamesmunns** [comment]
Oh, it looks like fixing `print.html` might be difficult, we currently [skip linkchecking print.html](https://github.com/rust-embedded/discovery/pull/142/files) in the embedded book, as mdbook does not render these links as expected.

@steveklabnik is the contents of `print.html` important in this context? If so, fixing this upstream in mdbook might be required.

**Comment by steveklabnik** [comment]
> Would you be up for supporting multiple versions of mdbook at the same time? I think this could solve our problem today, and it could make it easier to incrementally port books between versions.

I *think* this might be really helpful. I am not sure if @rust-lang/infra is happy with it though. If they are, I am! This would be super useful to make the ports incremental, and then we can revert to a single version once it's done.

> @steveklabnik is the contents of print.html important in this context? If so, fixing this upstream in mdbook might be required.

Yes; people print these to PDF and expect links to work and stuff. We should fix whatever the `print.html` issue is, for sure.

**Comment by steveklabnik** [comment]
ping @jamesmunns ! Is there anything I can do to help you get across the finish line here?

**Comment by jamesmunns** [comment]
Comments in https://github.com/rust-lang-nursery/edition-guide/pull/134#issuecomment-453204579 !

**Comment by jamesmunns** [comment]
Hard/manual rebased on https://github.com/rust-lang/rust/pull/57495. Also still blocked by https://github.com/rust-embedded/book/pull/94.

**Comment by jamesmunns** [comment]
@steveklabnik this is okay to `r+` after #57495 lands IMO! I don't think a rebase should be necessary, but happy to do if preferred.

**Comment by steveklabnik** [comment]
yes, r=me after that lands :)

**Comment by jamesmunns** [comment]
Pinging @steveklabnik, have I got a book for you.

**Comment by jamesmunns** [comment]
@steveklabnik ready for the `r+`!

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
I've rebased and squashed this; let's make sure travis passes, and then this looks good to go to me. @jamesmunns maybe triple check this?

**Comment by jamesmunns** [comment]
@steveklabnik it looks like the submodule rolled back in versions, updated with the current master branch. Otherwise LGTM!

**Comment by steveklabnik** [comment]
@bors: r+

Woo!

**Comment by steveklabnik** [comment]
@bors: retry

seems spurious and not related?

**Comment by pietroalbini** [comment]
Yep, it's spurious, we're trying to track it.

**Pull request merged**
